[Marcus Pang Yu Yang] iP

marcuspang opened the pull request:
## Dukegg
> “The bird fights its way out of the egg. The egg is the world. Whoever will be born must destroy a world.” – Hermann Hesse ([source](https://www.brainyquote.com/quotes/hermann_hesse_737841))

Duke🥚 frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ **_BLAZINGLY_** FAST to use

All you need to do is,

1. download it from [here](https://github.com/marcuspang/ip/releases/).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

<hr />

If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:


```java
public class Main {
    public static void main(String[] args) {
        new Duke("data/saveFile.txt").run();
    }
}
```

Commit:
Add Gradle support

marcuspang committed:
Implement while-loop to echo user's input

marcuspang committed:
Add task list feature

Refactored command execution process into instance methods, and initialized a chatbot instance inside the main function. Also, refactored long line into a separate constant and made use of StringBuilder instead of string concatenation where possible.

marcuspang committed:
Create Task class and add mask/unmask commands

marcuspang committed:
Update executeCommand to use switch statement instead of if-else

marcuspang committed:
Add Todo, Deadline, and Event classes

Also added *.class to .gitignore to prevent .class files from being committed

marcuspang committed:
Implement todo, deadline, and event commands

Removed NEW_LINE from LONG_LINE_WITH_NEW_LINE since all the lines will have a new line in the current implementation. Also, added a missing space for deadlines and events in the toString() method.

marcuspang committed:
Update test files

Also fixed number agreement in response text when adding tasks to the list

marcuspang committed:
Add DukeException

marcuspang committed:
Add toString method for DukeException

marcuspang committed:
Add errors and error handling for commands

marcuspang committed:
Fix checks for datetime field in event/deadline commands

marcuspang committed:
Add delete functionality

Also added initial capacity for task list of 100, as mentioned in Level-2

marcuspang committed:
Add enum for task types

Also removed unnecessary getTaskIcon method since it should return it's corresponding task icon

marcuspang committed:
Refactor add task logic

Moved the add task logic and response message to a separate method. Also added JavaDoc for the executeCommand and addTask methods

marcuspang committed:
Enforce code style

Mainly added javadoc for classes and class methods.

marcuspang committed:
Implement save/load tasks feature from local file

Fixed a bug where adding a task was printing the remove task message inside Duke.java. Also moved the multi-line operations to new lines to adhere to the code standards

marcuspang committed:
Convert String datetime fields to LocalDateTime

Also updated datetime variable names for Deadline and Event to be more semantic, i.e. deadline and timing

marcuspang committed:
Merge remote-tracking branch 'origin/branch-Level-7'

marcuspang committed:
Merge remote-tracking branch 'origin/branch-Level-8'

marcuspang committed:
Fix bug with storing datetime in storage file

marcuspang committed:
Create Ui, Parser, and TaskList classes and update Storage functionality

marcuspang committed:
Refactor classes to separate packages

marcuspang committed:
Merge branch 'add-gradle-support'

marcuspang committed:
Fix Gradle configuration for mainClassName

marcuspang committed:
Add tests for mark, todo, unmark commands and tasklist

marcuspang committed:
Add value to task type enum and refactored exceptions into separate files

marcuspang committed:
Add missing javadoc for command files

marcuspang committed:
Standardise comments to use American spelling and add missing period to comments

marcuspang committed:
Add find command and fix input validation for commands

Fixed input strings for description and datetime fields when splitting by checking with .trim().isEmpty(). Also refactored the printing of a list of tasks into a Ui method

marcuspang committed:
Add missing javadocs

marcuspang committed:
Merge branch 'branch-A-CodingStandard'

marcuspang committed:
Merge branch 'branch-Level-9'

marcuspang committed:
Add checkstyle and fix coding style violations

marcuspang committed:
Merge branch 'branch-A-CheckStyle'

marcuspang committed:
Add .vscode to .gitignore

ShummyOwnzYou commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,28 @@
+     * Indicates whether the chatbot should terminate after the command is executed.
+     *
+     * @return whether the chatbot should terminate.
+     */
```
Perhaps the @return tag could be "returns true if the chatbot should terminate, false otherwise"

marcuspang replied:
That's a good point, will update it

ShummyOwnzYou commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,83 @@
+     * @return New TaskList containing filtered tasks.
+     */
+    public TaskList filter(String query) {
+        return new TaskList(this.list.stream()
```
nice usage of stream

ShummyOwnzYou commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,143 @@
+     * Prints the goodbye message when the chatbot shuts down.
+     */
+    public void showGoodbye() {
+        System.out.println("     Bye. Hope to see you again soon!");
```
Perhaps you could consider using an indentation constant to append in front of all the `System.out.println` commands

marcuspang replied:
That's a good point, but since I have removed the indentation for level 10, I won't be addressing this issue.

ShummyOwnzYou reviewed (commented):
Overall, the code follows the coding standard.
No glaring mistakes that I can see. Good job!

marcuspang committed:
Add GUI and rename to Dukegg

marcuspang committed:
Update javadoc for isExit method and refactor app shutdown method to Dukegg

marcuspang committed:
Merge branch 'branch-Level-10'

marcuspang committed:
Rename package to dukegg

TheSoggy reviewed (commented):
Overall it LGTM! Coding standards were adhered to as far as I can see.

marcuspang committed:
Add assertions

To verify the correctness of an invariant in the code, we need to use assertions.

Consequently, we are adding assertion statements.

Added assertion statements to:
- getResponse method as the input can never be null
- createFileAndSubdirectoriesIfFileNotFound as the file must have been created after calling this method

This is to ensure during runtime that the methods listed behave as expected.

marcuspang committed:
Refactor invalid task exception

The error message was being defined twice for invalid index access in both MarkCommand and UnmarkCommand

To address this, I created a new exception class to handle the error message to be generated. I also found out that the DeleteCommand was missing the index out of bounds check, so it was also added to the DeleteCommand

A new exception class was created to standardise the error message and formatting to minimise the amount of refactoring required when changing the error message content or format.

marcuspang committed:
Create gradle.yml

marcuspang committed:
Fix GitHub workflow .yml file

marcuspang committed:
Merge pull request #1 from marcuspang/branch-A-Assertions

`A-Assertions` branch

marcuspang committed:
Merge pull request #2 from marcuspang/branch-A-CodeQuality

`A-CodeQuality` branch

marcuspang committed:
Merge pull request #3 from marcuspang/branch-A-CI

`A-CI` branch

marcuspang committed:
Remove unused imports

marcuspang committed:
Add tag to tasks and update storage methods

Add tag field to Task class and added new constructors to Task and its subclasses to make the creation of these tasks more flexible.

Storage will use constructors that require a tag to be passed as tags will always be stored in the storage file. If it is empty, it will be stored as null.

The commands will use contructors that do not require a tag as they will not be able to assign a tag to a new task.

marcuspang committed:
Add tag/untag commands

Implement the tagging and untagging features for the branch requirement.

Also added the necessary Ui methods and used a monospace font for the dialog text box.

marcuspang committed:
Add missing javadoc to new methods and remove unused imports

marcuspang committed:
Merge pull request #5 from marcuspang/branch-C-Tagging

`C-Tagging`

marcuspang committed:
Add styling to GUI using CSS and refactor DialogBox classes

CSS styles adopted from https://github.com/rebeccacxy/ip. DialogBox classes refactored into 2 separate dialog box classes, as the flip function was unnecessary since we can simply abstract the UI for the dialog boxes for chatbot and user into separate files.

marcuspang committed:
Add responsiveness to GUI

marcuspang committed:
Fix import order for command files

marcuspang committed:
Fix import order for test command files

marcuspang committed:
Add help command

marcuspang committed:
Merge pull request #6 from marcuspang/branch-A-BetterGui

`A-BetterGui` branch

marcuspang committed:
Replace emojis with 🥚

marcuspang committed:
Merge pull request #7 from marcuspang/branch-A-Personality

`A-Personality` Branch

marcuspang committed:
Replace emoji with 🥚 in InvalidTaskIndexException

marcuspang committed:
Add test for bye command

marcuspang committed:
Remove unused image files

marcuspang committed:
Add representative screenshot

marcuspang committed:
Update user guide in README.md and fix syntax for some commands

marcuspang committed:
Increase width of image in README.md
